File: heygen_video.py
"""Optional, authenticated HeyGen integration. No changes to the tutor's policy.

Provider contract: https://developers.heygen.com/reference/create-video.md
"""
from __future__ import annotations

import hashlib
import ipaddress
import json
import os
import re
import sqlite3
import time
from collections.abc import Callable
from typing import Literal
from urllib.parse import urlsplit
from uuid import UUID, uuid4

import httpx
from fastapi import APIRouter, Cookie, Depends, HTTPException, Query, Request, Response
from pydantic import BaseModel, ConfigDict, Field, field_validator

MAX_SCRIPT = 4000
ID_PATTERN = r"^[A-Za-z0-9_-]{1,200}$"


class VideoRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)
    sessionId: str = Field(min_length=1, max_length=200)
    requestId: UUID
    script: str = Field(min_length=1, max_length=MAX_SCRIPT)
    avatarId: str = Field(pattern=ID_PATTERN)
    voiceId: str = Field(default="", max_length=200)
    reviewed: Literal[True]

    @field_validator("voiceId")
    @classmethod
    def valid_voice(cls, value: str) -> str:
        if value and not re.fullmatch(ID_PATTERN, value):
            raise ValueError("Use a HeyGen voice ID, not a URL or API key.")
        return value


def public_https(value: object) -> str | None:
    """Only allow public HTTPS delivery links; never fetch them server-side."""
    if not isinstance(value, str) or len(value) > 8000:
        return None
    try:
        parsed = urlsplit(value)
        host = (parsed.hostname or "").lower().rstrip(".")
        if parsed.scheme != "https" or parsed.username or parsed.password or parsed.port not in (None, 443):
            return None
        if not host or "." not in host or host.endswith((".localhost", ".local", ".internal", ".test", ".lan")):
            return None
        try:
            ipaddress.ip_address(host)
        except ValueError:
            return value
    except ValueError:
        pass
    return None


class HeyGenClient:
    """Fixed provider endpoint, bounded responses, sanitized errors, no auto-retries."""

    def __init__(self, api_key: str | None = None, transport: httpx.BaseTransport | None = None):
        self.api_key = (api_key if api_key is not None else os.getenv("HEYGEN_API_KEY", "")).strip()
        self.transport = transport

    def request(self, method: str, path: str, *, payload: dict | None = None,
                params: dict | None = None, request_key: str | None = None) -> dict:
        if not self.api_key:
            raise HTTPException(503, "Add HEYGEN_API_KEY to your local .env, save it, and restart the Python app.")
        headers = {"X-Api-Key": self.api_key, "Accept": "application/json"}
        if request_key:
            headers["Idempotency-Key"] = request_key
        try:
            with httpx.Client(timeout=40, follow_redirects=False, transport=self.transport) as client:
                with client.stream(method, "https://api.heygen.com" + path, headers=headers, json=payload, params=params) as response:
                    if not response.is_success:
                        messages = {
                            400: "HeyGen rejected the script, avatar or voice settings. Check the selected IDs and API plan.",
                            401: "HeyGen authentication failed. Check that your saved key is a valid HeyGen API key.",
                            402: "HeyGen API credits are insufficient. Check API billing (separate from website credits).",
                            403: "HeyGen access denied. Check API permissions and access to this avatar or voice.",
                            404: "HeyGen could not find this resource. Check the avatar, voice or video ID.",
                            409: "HeyGen is still handling this request. Wait briefly, then retry the same submission.",
                            429: "HeyGen rate limit or quota reached. Wait before checking again; do not generate another video.",
                        }
                        code = response.status_code
                        raise HTTPException(code if code in (409, 429) else 502, messages.get(code, f"HeyGen is unavailable (HTTP {code}). Check its dashboard before retrying."))
                    chunks = bytearray()
                    for chunk in response.iter_bytes():
                        chunks.extend(chunk)
                        if len(chunks) > 2 * 1024 * 1024:
                            raise HTTPException(502, "HeyGen returned an oversized response.")
                    result = json.loads(chunks)
        except httpx.TimeoutException as exc:
            raise HTTPException(504, "HeyGen timed out. A video may still have been submitted. Retry the same request, not a new video.") from exc
        except httpx.RequestError as exc:
            raise HTTPException(502, "Could not reach HeyGen. Check its dashboard before creating another video.") from exc
        except (ValueError, UnicodeError) as exc:
            raise HTTPException(502, "HeyGen returned an unreadable response.") from exc
        if not isinstance(result, dict) or result.get("error"):
            raise HTTPException(502, "HeyGen returned an error. Check your API account and selected settings.")
        return result

    def avatars(self, token: str = "") -> dict:
        params = {"ownership": "public", "limit": 50}
        if token:
            params["token"] = token
        result = self.request("GET", "/v3/avatars/looks", params=params)
        items = result.get("data")
        if not isinstance(items, list):
            raise HTTPException(502, "HeyGen returned an invalid avatar list.")
        avatars = []
        for item in items[:50]:
            if not isinstance(item, dict) or not re.fullmatch(ID_PATTERN, str(item.get("id", ""))):
                continue
            if item.get("status") not in (None, "completed"):
                continue
            engines = item.get("supported_api_engines")
            if isinstance(engines, list) and "avatar_iv" not in engines:
                continue
            avatars.append({"id": item["id"], "name": str(item.get("name", "Avatar"))[:200]})
        token = result.get("next_token")
        return {"avatars": avatars, "nextToken": token if result.get("has_more") and isinstance(token, str) and len(token) <= 2048 else None}

    def create(self, value: VideoRequest, request_key: str) -> str:
        payload = {"type": "avatar", "avatar_id": value.avatarId, "script": value.script,
                   "title": "Math coach explanation", "resolution": "720p", "aspect_ratio": "16:9"}
        if value.voiceId:
            payload["voice_id"] = value.voiceId
        result = self.request("POST", "/v3/videos", payload=payload, request_key=request_key)
        data = result.get("data")
        video_id = data.get("video_id") if isinstance(data, dict) else None
        if not isinstance(video_id, str) or not re.fullmatch(ID_PATTERN, video_id):
            raise HTTPException(502, "HeyGen did not return a video ID. Check its dashboard before trying a new request.")
        return video_id

    def status(self, video_id: str) -> dict:
        if not re.fullmatch(ID_PATTERN, video_id):
            raise HTTPException(502, "Invalid stored video ID.")
        data = self.request("GET", f"/v3/videos/{video_id}").get("data")
        if not isinstance(data, dict) or data.get("id") != video_id:
            raise HTTPException(502, "HeyGen returned invalid video details.")
        state = data.get("status")
        if state not in ("pending", "processing", "completed", "failed"):
            raise HTTPException(502, "HeyGen returned an unknown video status. Check again later.")
        url = public_https(data.get("video_url")) if state == "completed" else None
        if state == "completed" and not url:
            raise HTTPException(502, "The video is complete but its delivery link is unavailable. Check again later.")
        return {"status": state, "videoUrl": url,
                "message": "HeyGen could not render this video. Check its dashboard for details before generating another." if state == "failed" else None}


def create_video_router(authenticate: Callable, connect: Callable[[], sqlite3.Connection],
                        provider_factory: Callable[[], HeyGenClient] = HeyGenClient) -> APIRouter:
    """Inject existing authentication/storage instead of importing or changing the tutor."""
    with connect() as db:
        db.execute("""CREATE TABLE IF NOT EXISTS heygen_video_jobs (
            id TEXT PRIMARY KEY, student_id INTEGER NOT NULL, session_id TEXT NOT NULL,
            request_id TEXT NOT NULL, fingerprint TEXT NOT NULL, video_id TEXT,
            status TEXT NOT NULL, created_at REAL NOT NULL, submitted_at REAL NOT NULL,
            UNIQUE(student_id, request_id)
        )""")

    def student(request: Request, response: Response, math_tutor_session: str | None = Cookie(default=None)):
        response.headers["Cache-Control"] = "no-store"
        origin = request.headers.get("origin")
        if request.headers.get("sec-fetch-site") == "cross-site" or (origin and origin != str(request.base_url).rstrip("/")):
            raise HTTPException(403, "Cross-origin video requests are not allowed.")
        return authenticate(math_tutor_session)

    router = APIRouter(prefix="/api/heygen", tags=["Tutor videos"])

    def owned_session(db, session_id, student_id):
        if db.execute("SELECT id FROM practice_sessions WHERE id = ? AND student_id = ?", (session_id, student_id)).fetchone() is None:
            raise HTTPException(404, "Practice session not found.")

    def public_job(row):
        return {"id": row["id"], "sessionId": row["session_id"], "status": row["status"], "createdAt": row["created_at"], "videoUrl": None}

    @router.get("/config")
    def config(learner=Depends(student)):
        return {"configured": bool(os.getenv("HEYGEN_API_KEY", "").strip()), "maxScriptLength": MAX_SCRIPT,
                "defaultAvatarId": os.getenv("HEYGEN_AVATAR_ID", ""), "defaultVoiceId": os.getenv("HEYGEN_VOICE_ID", "")}

    @router.get("/avatars")
    def avatars(token: str = Query(default="", max_length=2048), learner=Depends(student)):
        return provider_factory().avatars(token)

    @router.get("/videos")
    def videos(sessionId: str = Query(min_length=1, max_length=200), learner=Depends(student)):
        with connect() as db:
            owned_session(db, sessionId, learner["id"])
            rows = db.execute("SELECT * FROM heygen_video_jobs WHERE student_id = ? AND session_id = ? ORDER BY created_at DESC LIMIT 10", (learner["id"], sessionId)).fetchall()
        return {"videos": [public_job(row) for row in rows]}

    @router.post("/videos", status_code=202)
    def generate(value: VideoRequest, learner=Depends(student)):
        # The provider's idempotency window is 24 hours; never blindly retry later.
        fingerprint = hashlib.sha256(json.dumps(value.model_dump(mode="json"), sort_keys=True).encode()).hexdigest()
        now = time.time()
        with connect() as db:
            db.execute("BEGIN IMMEDIATE")
            owned_session(db, value.sessionId, learner["id"])
            previous = db.execute("SELECT * FROM heygen_video_jobs WHERE student_id = ? AND request_id = ?", (learner["id"], str(value.requestId))).fetchone()
            if previous:
                if previous["fingerprint"] != fingerprint:
                    raise HTTPException(409, "This submission ID belongs to a different script or selection. Start a new reviewed draft.")
                if previous["video_id"]:
                    return public_job(previous)
                if now - previous["created_at"] >= 23 * 3600:
                    raise HTTPException(409, "The safe retry window has expired. Check HeyGen's dashboard before creating a new video.")
                if previous["status"] == "submitting" and now - previous["submitted_at"] < 90:
                    raise HTTPException(409, "This video is being submitted. Wait before retrying the same request.")
                job_id = previous["id"]
                db.execute("UPDATE heygen_video_jobs SET status = 'submitting', submitted_at = ? WHERE id = ?", (now, job_id))
            else:
                # Keep this local prototype from accidentally creating expensive batches.
                active = db.execute("SELECT COUNT(*) FROM heygen_video_jobs WHERE student_id = ? AND status IN ('submitting', 'pending', 'processing', 'submission_unknown') AND created_at > ?", (learner["id"], now - 86400)).fetchone()[0]
                if active >= 3:
                    raise HTTPException(429, "You already have three active or unconfirmed videos. Check their status or the HeyGen dashboard first.")
                job_id = str(uuid4())
                db.execute("INSERT INTO heygen_video_jobs (id, student_id, session_id, request_id, fingerprint, status, created_at, submitted_at) VALUES (?, ?, ?, ?, ?, 'submitting', ?, ?)", (job_id, learner["id"], value.sessionId, str(value.requestId), fingerprint, now, now))
        try:
            video_id = provider_factory().create(value, f"edume:{learner['id']}:{value.requestId}")
        except HTTPException as exc:
            with connect() as db:
                db.execute("UPDATE heygen_video_jobs SET status = ? WHERE id = ?", ("failed" if exc.status_code == 503 else "submission_unknown", job_id))
            raise
        with connect() as db:
            db.execute("UPDATE heygen_video_jobs SET video_id = ?, status = 'pending' WHERE id = ?", (video_id, job_id))
            row = db.execute("SELECT * FROM heygen_video_jobs WHERE id = ?", (job_id,)).fetchone()
        return public_job(row)

    @router.get("/videos/{job_id}")
    def video(job_id: UUID, learner=Depends(student)):
        with connect() as db:
            row = db.execute("SELECT * FROM heygen_video_jobs WHERE id = ? AND student_id = ?", (str(job_id), learner["id"])).fetchone()
        if row is None:
            raise HTTPException(404, "Video not found.")
        result = public_job(row)
        if not row["video_id"]:
            result["message"] = "Submission is unconfirmed. Check HeyGen's dashboard; retry only the same unchanged draft to avoid duplicate charges."
            return result
        details = provider_factory().status(row["video_id"])
        with connect() as db:
            db.execute("UPDATE heygen_video_jobs SET status = ? WHERE id = ?", (details["status"], str(job_id)))
        return {**result, **details}

    return router

File: static/heygen-video.js
// One integration point: tutor output in, reviewed narration out. Never auto-generate.
export function tutorScript(tutor) {
  if (!tutor) return '';
  return [tutor.prompt, tutor.workedExample?.title, ...(tutor.workedExample?.steps || []), tutor.workedExample?.handoff, tutor.visualCue]
    .filter(value => typeof value === 'string' && value.trim()).join('\n\n');
}

export function setupHeygenVideo({ getSessionId }) {
  const $ = id => document.getElementById(id), dialog = $('heygen-dialog');
  let source = '', draftSession = null, generation = 0, pollTimer = null, pollCount = 0;
  let busy = false, config = null, submitted = null, activeJob = null, nextToken = null;
  const pendingStates = new Set(['pending', 'processing', 'submitting']);
  function message(text) { $('heygen-status').textContent = text; }
  function refresh() {
    const script = $('heygen-script').value.trim();
    $('heygen-count').textContent = `${$('heygen-script').value.length} / 4000 characters`;
    $('heygen-generate').disabled = busy || !config?.configured || !script || script.length > 4000 || !$('heygen-avatar').value || !$('heygen-consent').checked || Boolean(activeJob);
    for (const id of ['heygen-script', 'heygen-avatar', 'heygen-voice', 'heygen-consent']) $(id).disabled = busy || Boolean(submitted);
    $('heygen-load-avatars').disabled = busy || !config?.configured || Boolean(submitted);
    $('heygen-more-avatars').disabled = busy || !nextToken || Boolean(submitted);
    $('heygen-refresh').disabled = busy || !activeJob;
    $('heygen-new-draft').disabled = busy;
    $('heygen-generate').textContent = submitted && !activeJob ? 'Retry same submission' : 'Generate video · uses credits';
  }
  function clearPlayer() {
    $('heygen-player').pause(); $('heygen-player').removeAttribute('src'); $('heygen-player').hidden = true;
    $('heygen-download').removeAttribute('href'); $('heygen-download').hidden = true;
  }
  function stopPolling() { clearTimeout(pollTimer); pollTimer = null; }
  async function api(path, options = {}) {
    const response = await fetch(`/api/heygen${path}`, { ...options, headers: { 'Content-Type': 'application/json', ...options.headers }, signal: AbortSignal.timeout(55000) });
    let value;
    try { value = await response.json(); } catch { throw new Error('The server returned an unreadable response. Check its connection.'); }
    if (!response.ok) throw new Error(typeof value.detail === 'string' ? value.detail : 'Invalid video request. Review the script and settings.');
    return value;
  }
  async function operation(action) {
    if (busy) return;
    const token = generation; busy = true; refresh();
    try { await action(() => generation === token && dialog.open); }
    catch (error) {
      if (generation === token) message(error.name === 'TimeoutError' || error instanceof TypeError ? 'Connection interrupted. If submitting, retry the same unchanged draft or check HeyGen before generating another video.' : error.message);
    } finally { if (generation === token) { busy = false; refresh(); } }
  }
  function showJob(job) {
    activeJob = job.id; $('heygen-job-id').textContent = `Local job: ${job.id}`;
    clearPlayer();
    if (job.status === 'completed' && job.videoUrl) {
      const url = new URL(job.videoUrl);
      if (url.protocol !== 'https:' || url.username || url.password) throw new Error('Invalid video delivery link.');
      $('heygen-player').src = url.href; $('heygen-player').hidden = false;
      $('heygen-download').href = url.href; $('heygen-download').hidden = false;
      message('Video ready. AI avatar narration—check the spoken math before sharing. Delivery links may expire; Check status refreshes them.');
    } else message(job.message || ({ pending: 'Video queued with HeyGen…', processing: 'HeyGen is rendering your video…', submitting: 'Submitting to HeyGen…', failed: 'Generation failed. Check HeyGen before trying again.', submission_unknown: 'Submission outcome is unknown. Check HeyGen before generating another video.' }[job.status] || 'Check status for an update.'));
    if (pendingStates.has(job.status) && dialog.open && pollCount < 75) {
      stopPolling(); pollTimer = setTimeout(() => { pollCount++; checkStatus(); }, 8000);
    } else if (pendingStates.has(job.status)) message('Still rendering. Automatic checks paused after 10 minutes; use Check status later.');
    refresh();
  }
  function checkStatus() {
    if (!activeJob || busy) return;
    stopPolling(); const jobId = activeJob;
    operation(async current => { const job = await api(`/videos/${encodeURIComponent(jobId)}`); if (current() && activeJob === jobId) showJob(job); });
  }
  function renderHistory(videos) {
    $('heygen-history').replaceChildren();
    for (const job of videos) {
      const button = document.createElement('button'); button.type = 'button'; button.className = 'heygen-history-item';
      button.textContent = `${new Date(job.createdAt * 1000).toLocaleTimeString()} · ${job.status.replaceAll('_', ' ')} · ${job.id.slice(0, 8)}`;
      button.addEventListener('click', () => { if (busy) return; stopPolling(); activeJob = job.id; pollCount = 0; checkStatus(); });
      $('heygen-history').append(button);
    }
  }
  async function loadHistory(current) {
    const history = await api(`/videos?sessionId=${encodeURIComponent(draftSession)}`);
    if (current()) renderHistory(history.videos);
  }
  function blankDraft() {
    stopPolling(); submitted = null; activeJob = null; pollCount = 0;
    $('heygen-script').value = source; $('heygen-consent').checked = false; $('heygen-job-id').textContent = ''; clearPlayer(); refresh();
  }
  $('heygen-open').addEventListener('click', () => {
    if (!source || !getSessionId()) return;
    const changed = draftSession !== getSessionId(); draftSession = getSessionId();
    dialog.showModal();
    if (changed || !submitted && !activeJob) blankDraft();
    operation(async current => {
      const settings = await api('/config'); if (!current()) return;
      config = settings;
      if (!$('heygen-avatar').value && settings.defaultAvatarId) {
        const option = document.createElement('option'); option.value = settings.defaultAvatarId; option.textContent = 'Configured avatar'; $('heygen-avatar').append(option); $('heygen-avatar').value = settings.defaultAvatarId;
      }
      if (!submitted && !$('heygen-voice').value) $('heygen-voice').value = settings.defaultVoiceId;
      message(settings.configured ? 'Review the narration, load an avatar, and confirm before spending credits.' : 'Add HEYGEN_API_KEY to your local .env, save, and restart the Python app.');
      await loadHistory(current);
      if (current() && activeJob) {
        const jobId = activeJob;
        const job = await api(`/videos/${encodeURIComponent(jobId)}`);
        if (current() && activeJob === jobId) showJob(job);
      }
    });
  });
  function loadAvatars(more = false) {
    operation(async current => {
      message('Loading public HeyGen avatars…');
      const data = await api(`/avatars${more && nextToken ? `?token=${encodeURIComponent(nextToken)}` : ''}`);
      if (!current()) return;
      if (!more) $('heygen-avatar').replaceChildren();
      for (const avatar of data.avatars) {
        const option = document.createElement('option'); option.value = avatar.id; option.textContent = avatar.name; $('heygen-avatar').append(option);
      }
      nextToken = data.nextToken;
      message(data.avatars.length ? 'Choose a public avatar. Its default voice is used unless you supply a HeyGen voice ID.' : 'No compatible avatars on this page. Try More avatars or check your HeyGen account.');
    });
  }
  $('heygen-load-avatars').addEventListener('click', () => loadAvatars());
  $('heygen-more-avatars').addEventListener('click', () => loadAvatars(true));
  for (const id of ['heygen-script', 'heygen-avatar', 'heygen-voice']) $(id).addEventListener('input', () => { $('heygen-consent').checked = false; refresh(); });
  $('heygen-consent').addEventListener('change', refresh);
  $('heygen-generate').addEventListener('click', () => {
    if ($('heygen-generate').disabled || !draftSession || draftSession !== getSessionId()) return;
    operation(async current => {
      // Preserve the exact draft + request ID across errors. Retrying must not charge twice.
      submitted ||= { sessionId: draftSession, requestId: crypto.randomUUID(), script: $('heygen-script').value.trim(), avatarId: $('heygen-avatar').value, voiceId: $('heygen-voice').value.trim(), reviewed: true };
      message('Submitting once to HeyGen. This uses API credits and may take several minutes…'); refresh();
      const job = await api('/videos', { method: 'POST', body: JSON.stringify(submitted) });
      if (!current()) return;
      showJob(job); await loadHistory(current);
    });
  });
  $('heygen-refresh').addEventListener('click', () => { pollCount = 0; checkStatus(); });
  $('heygen-new-draft').addEventListener('click', () => {
    if (busy) return;
    if ((submitted || activeJob) && !confirm('A new draft can create another paid video. Existing generation is NOT cancelled. Check its status or HeyGen dashboard first. Continue?')) return;
    blankDraft(); message('New draft ready. Review pronunciation and confirm before generating.');
  });
  $('heygen-close').addEventListener('click', () => dialog.close());
  dialog.addEventListener('close', () => { generation++; busy = false; stopPolling(); clearPlayer(); });
  window.addEventListener('pagehide', stopPolling);
  refresh();
  return {
    setGuidance(tutor) { source = tutorScript(tutor); $('heygen-open').disabled = !source || !getSessionId(); },
    reset() {
      generation++; stopPolling(); busy = false; source = ''; submitted = null; activeJob = null; config = null; draftSession = null;
      $('heygen-open').disabled = true; $('heygen-history').replaceChildren(); $('heygen-script').value = ''; clearPlayer();
      if (dialog.open) dialog.close(); refresh();
    }
  };
}